### PR TITLE
Use QImage instead of Pillow

### PR DESCRIPTION
**Pull request by leiserfg**
**Please check if the PR fulfills these requirements:**

- [ ] The commit message follows our guidelines.
- For bug fixes and features:
    - [X ] You tested the changes.
    - [ ] You updated the docs or changelog.

Related issue (if applicable): #38 

**What kind of change does this PR introduce?**

Replaces Pillow with QImage

**Does this PR introduce a breaking change?**

## New feature or change ##

**What is the current behavior?** 
The add-on requires Pillow in order to work

**What is the new behavior?**
It works with QImage instead

**Other information**

### CODE REVIEW

**Comment by leiserfg** [comment]
Sorry, the PR also includes the changes for the other one but I needed to do so in order to test it on Krita 4.3.
There is a similar bug with COA but I didn't fix that one because I didn't wanna keep adding more changes to this one.

**Comment by NathanLovato** [comment]
Thanks much, I've asked Razvan to test and review it.

**Comment by razcore-art** [comment]
Thanks for your PR, I don't have much experience with QT myself and so I didn't even think about QIamge/QPainter although I'm sure it's used even in the Krita's Python engine :). Very much appreciated.

I left you some reviews, they're just about formatting, the rest is peachy perfect. Thanks again.

**Comment by leiserfg** [comment]
Sorry @razcore-art, where did you leave your reviews?

**Comment by NathanLovato** [comment]
@razcore-art Did you confirm your review comments? I'm not seeing any comments.

**Comment by razcore-art** [comment]
Ups, sorry guys, I was still doing tests on the PR and forgot to submit the review.

File: krita_batch_exporter/Infrastructure.py [hunk 1]
```diff
@@ -1,67 +1,44 @@
...
-        temp_doc.close()
 
-    return img
-
-
-def toJPEG(img):
-    newImg = Image.new("RGBA", img.size, 4 * (255,))
-    newImg.alpha_composite(img)
-    return newImg.convert("RGB")
+def expand_and_format(img, margin=0, is_jpg=False):
```
[the remark this answers is not in the document]
Comment (leiserfg) [reply krita_batch_exporter/Infrastructure.py:26]: Sorry, I use pep8 so I did it automatically.

**Comment by razcore-art** [comment]
After playing a bit with it, looks like we have some issues to deal with. I'm gonna give you a link to a *.kra file I worked with with the `Pillow` version for a tutorial on GDQuest website.

https://www.dropbox.com/s/vvlu3v5omhhb1m1/world-map-raw.zip?dl=0

You can experiment with it, it's a bit more complex than just regular layers. It has transparency masks etc. You can compare the results with these expected images: https://www.dropbox.com/sh/p87vupx033qxbt7/AAAzGSuDetBHh4MnUdDDaVWza?dl=0.

As it stands there are these issues:
- the QImage/QPainter approach produces sub-par quality, at least with the above `kra` file
- in some cases, seems that using transparency masks produces images with extra margins. This might be due to the use of `wnode.node.thumbnail()`. I looked at the documentation, but it isn't clear for me what it does.

The reason why we ended up using `Pillow` in the first place is because:

1. It was faster than doing the same with Krita actions
2. Some Krita actions (like scale) were run in parallel and the output was saved before the computation was finished so we ended up with empty images. This last issue might have been fixed, I haven't checked.

It seems that, although `QImage` has this `smoothResize()` function, it doesn't produce very good quality, or it might, again be due to the use of `wnode.node.thumbnail()`.

Here's an example of an image produced with this PR:
gradient-discrete

Expected result & quality:
gradient-discrete

Notice the 2 things:
1. The quality of the image itself isn't as good
2. The size of the image is wrong, there's extra margin on the right & bottom sides.

**Comment by leiserfg** [comment]
I know Pillow is faster than using krita and that it has a good algorithm for interpolating (Lanczos) I'm just trying to remove the dependency to make easier integrating the plugin into krita itself, I will try to resolve the issue with the margin first.

**Comment by razcore-art** [comment]
Yeah, I'm not entirely sure about that `node.thumbnail()` does what we want it to do. I think it's better to rely on `node.projectionPixelData()`, we'll just have to see how to work with `QByteArray` and `QImage` if we want to get to the bottom of the bad quality.

**Comment by leiserfg** [comment]
The problem is not with thumbnail, that's what krita uses to draw the thumbnails on the layer view and the result is fine, my original idea was to use it directly with the required scale, I just didn't have a complex enough .kra to test my idea. That will provide a good result for down-scaling at least (thumbnail does not work for sizes over the layer size).

**Comment by leiserfg** [comment]
Maybe the margin error is because QPainter drawing a argb image over a rgb one, I will experiment with it.

**Comment by NathanLovato** [comment]
As long as the output quality, when downscaling, is as good as with Pillow, losing some speed is an okay tradeoff for removing the installation issue and making it possible to merge into Krita.

As you said, upscaling on export isn't too useful, and, as far as I'm concerned, it's not a supported use-case for this add-on. What we want is to allow people to export crisp images for games, the web, comics, designs... all that with a lot of flexibility in how they organize their Krita files.

**Comment by leiserfg** [comment]
The issue was with the `thumbnail` so I rolled back that part but keeping QImage, I also increased the quality for `jpg` to reduce artifacts.

File: krita_batch_exporter/Infrastructure.py [hunk 2]
```diff
@@ -19,13 +19,48 @@ def nodeToImage(wnode):
...
+        pixel_data = wnode.node.projectionPixelData(x, y, w, h).data()
+        temp_layer.setPixelData(pixel_data, x, y, w, h)
+
+        temp_doc.setColorSpace("RGBA", "U8", SRGB_PROFILE)
+        assert temp_doc.colorModel() == "RGBA" and temp_doc.colorDepth() == "U8"
+        assert temp_layer.colorModel() == "RGBA" and temp_layer.colorDepth() == "U8"
+
+        pixel_data = temp_layer.projectionPixelData(x, y, w, h).data()
+
+        temp_doc.close()
```
Comment (razcore-art) [inline krita_batch_exporter/Infrastructure.py]: Seems like all of this stuff can be replaced with:
```suggestion
        temp_node = wnode.node.duplicate()
        temp_node.setColorSpace(wnode.node.colorModel(), wnode.node.colorDepth(), SRGB_PROFILE)
        assert temp_node.colorModel() == "RGBA" and temp_node.colorDepth() == "U8"
        pixel_data = temp_node.projectionPixelData(x, y, w, h).data()
        del temp_node
```

Can you please confirm at your end that it doesn't break anything? If all's good looks like it's ready to merge.

Comment (leiserfg) [reply krita_batch_exporter/Infrastructure.py]: Sure, I can test it today after working hours.

**Comment by leiserfg** [comment]
I test it and it worked fine for me.

**Comment by razcore-art** [comment]
Done! Thanks a lot. If you have any other ideas feel free to open issues/PRs :)

**Comment by NathanLovato** [comment]
Great, thanks a lot for your contribution! Batch exporter should soon be part of Krita :)

**Pull request merged**